### PR TITLE
Detect how to connect to the box

### PR DESCRIPTION
**Pull request by gmarty**

### CODE REVIEW

File: app/js/lib/foxbox.js [hunk 1]
```diff
@@ -135,7 +179,8 @@ export default class Foxbox extends Service {
           // Check if we have a recent registry.
           const now = Math.floor(Date.now() / 1000);
           if ((now - box.timestamp) < 60) {
-            settings.hostname = box.hostname || box.local_ip;
+            settings.localHostname = box.local_ip;
+            settings.tunnelHostname = box.tunnel_url;
```
Comment (arcturus) [inline app/js/lib/foxbox.js]: The tunnel_url can be optional, is ok saving it to null or undefined.

File: app/js/lib/foxbox.js [hunk 2]
```diff
@@ -173,6 +219,66 @@ export default class Foxbox extends Service {
...
+
+    return Promise.resolve();
+  }
+
+  /**
+   * Ping the box to detect whether we connect locally or remotely.
+   */
+  pingBox() {
+    Promise.all([
+        ping(`${this.localOrigin}/ping`),
```
Comment (arcturus) [inline app/js/lib/foxbox.js]: Do we have the ping urls?

File: app/js/lib/foxbox.js [hunk 3]
```diff
@@ -173,6 +219,66 @@ export default class Foxbox extends Service {
...
+      .then(states => {
+        const connectedLocally = states[0];
+        const connectedToTunnel = states[1];
+
+        _connected = true;
+
+        if (connectedLocally) {
+          _local = true;
+        } else if (connectedToTunnel) {
+          _local = false;
```
Comment (arcturus) [inline app/js/lib/foxbox.js]: nit: _local = !!connectedLocally;

File: app/js/lib/foxbox.js [hunk 4]
```diff
@@ -173,6 +219,66 @@ export default class Foxbox extends Service {
...
+      // We also ping the box every few minutes to make sure it's still there.
+      _onlineInterval = setInterval(this.pingBox.bind(this),
+        settings.onlineCheckingLongInterval);
+    } else {
+      // If the Network Information API is not implemented, fallback to polling.
+      _onlineInterval = setInterval(this.pingBox.bind(this),
+        settings.onlineCheckingInterval);
+    }
+
+    this.pingBox();
```
Comment (arcturus) [inline app/js/lib/foxbox.js]: Don't we want to waint until the pingBox method finished? Should we return it?

File: app/js/lib/foxbox.js [hunk 5]
```diff
@@ -173,6 +219,66 @@ export default class Foxbox extends Service {
...
+   */
+  _initNetwork() {
+    window.addEventListener('online', this.pingBox.bind(this));
+    window.addEventListener('offline', this.pingBox.bind(this));
+
+    if ('connection' in navigator && 'onchange' in navigator.connection) {
+      navigator.connection.addEventListener('change', this.pingBox.bind(this));
+
+      // We also ping the box every few minutes to make sure it's still there.
+      _onlineInterval = setInterval(this.pingBox.bind(this),
```
Comment (arcturus) [inline app/js/lib/foxbox.js]: Why are we pinging here the box if we have the navigator.connection api?

Comment (gmarty) [inline app/js/lib/foxbox.js]: To detect if the box gets disconnected from the internet (unplugged, turned off, power outage...).

Comment (arcturus) [inline app/js/lib/foxbox.js]: Oh great, sorry I missed the 'long' interval part :)

**Comment by arcturus** [comment]
rplused!

Great job @gmarty !

**Pull request merged**